packages/core/src/page.js

```diff
@@ -49,17 +49,39 @@ export class Page {
   }
 
   // Go to a URL and wait for navigation to occur
-  async goto(url, { waitUntil = 'load' } = {}) {
+  async goto(url, { waitUntil = 'load', cookies } = {}) {
     this.log.debug(`Navigate to: ${url}`, this.meta);
 
     let navigate = async () => {
+      const mergeCookies = (userPassedCookie, autoCapturedCookie) => {
+        const mergedCookies = [...userPassedCookie, ...autoCapturedCookie];
+        const uniqueCookies = [];
+        const names = new Set();
+
+        for (const cookie of mergedCookies) {
+          if (!names.has(cookie.name)) {
+            uniqueCookies.push(cookie);
+            names.add(cookie.name);
+          }
+        }
+
+        return uniqueCookies;
+      };
+
+      const userPassedCookie = this.session.browser.cookies;
       // set cookies before navigation so we can default the domain to this hostname
-      if (this.session.browser.cookies.length) {
+      if (userPassedCookie.length || cookies) {
         let defaultDomain = hostname(url);
 
+        if (userPassedCookie.length > 0 && cookies) {
```

Comment:
Pass both userPassedCookie and cookies in the method suggested above and handle all conditions there, so all cookie related stuff will be in that method and easy to debug in future.